src/libsystemd/sd-event/sd-event.c

```diff
@@ -2242,188 +2242,195 @@ _public_ int sd_event_source_get_enabled(sd_event_source *s, int *m) {
+                }
 
-                        s->enabled = m;
+                break;
 
-                        if (EVENT_SOURCE_WATCH_PIDFD(s)) {
-                                /* yes, we have pidfd */
+        case SOURCE_EXIT:
+                prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
+                break;
 
-                                r = source_child_pidfd_register(s, s->enabled);
-                                if (r < 0) {
-                                        s->enabled = SD_EVENT_OFF;
-                                        s->event->n_enabled_child_sources--;
-                                        return r;
-                                }
-                        } else {
-                                /* no pidfd, or something other to watch for than WEXITED */
+        case SOURCE_DEFER:
+        case SOURCE_POST:
+        case SOURCE_INOTIFY:
+                break;
+        default:
+                assert_not_reached("Wut? I shouldn't exist.");
+        }
 
-                                r = event_make_signal_data(s->event, SIGCHLD, NULL);
-                                if (r < 0) {
-                                        s->enabled = SD_EVENT_OFF;
-                                        s->event->n_enabled_child_sources--;
-                                        event_gc_signal_data(s->event, &s->priority, SIGCHLD);
-                                        return r;
-                                }
-                        }
+        source_prioq_reshuffle(s);
 
-                        break;
+        return 0;
+}
 
-                case SOURCE_EXIT:
-                        s->enabled = m;
-                        prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
-                        break;
+_public_ int sd_event_source_set_enabled(sd_event_source *s, int m) {
+        int r;
 
-                case SOURCE_DEFER:
-                case SOURCE_POST:
-                case SOURCE_INOTIFY:
-                        s->enabled = m;
-                        break;
+        assert_return(s, -EINVAL);
+        assert_return(IN_SET(m, SD_EVENT_OFF, SD_EVENT_ON, SD_EVENT_ONESHOT), -EINVAL);
+        assert_return(!event_pid_changed(s->event), -ECHILD);
 
-                default:
-                        assert_not_reached("Wut? I shouldn't exist.");
-                }
-        }
+        /* If we are dead anyway, we are fine with turning off
+         * sources, but everything else needs to fail. */
+        if (s->event->state == SD_EVENT_FINISHED)
+                return m == SD_EVENT_OFF ? 0 : -ESTALE;
 
-        if (s->pending)
-                prioq_reshuffle(s->event->pending, s, &s->pending_index);
+        if (s->enabled == m)
+                return 0;
 
-        if (s->prepare)
-                prioq_reshuffle(s->event->prepare, s, &s->prepare_index);
+        s->enabled = m;
+        r = m == SD_EVENT_OFF ? source_disable(s) : source_enable(s, m);
+        if (r < 0)
```

Comment:
I played around with this in the previous PR.

I found a problem with setting s->enabled before and after the line following because it needed to be conditionally modified in the body of the factored out functions. I just left the modification of that field to those functions rather than pulling it out.

I haven't yet played around with this commit so the situation might not be the same, I'll try and have a look tomorrow, ;)

Reply:
> I just left the modification of that field to those functions rather than pulling it out

However, this is exactly what I want to avoid. `source->enabled` can't be altered by newly introduced low-level functions because we don't want to alter state of event source specified by user when changing its low-level state (e.g. whether it is monitored by epoll instance).